### ISSUE TITLE
Noticeable flicker/delay in icons when working with files

### ISSUE THREAD

**Issue by sgerace** [jisaacks/GitGutter#340]
So I got the latest update today (1.4.0) and I'm seeing a very noticable flicker/delay from when I update the code and when the gitgutter icons are refreshed. It looks like the state of the lines is not being remembered and instead all of the lines are cleared and refreshed. I can't say for certain that this was caused by the 1.4.0 update, however, I feel like I would have noticed this if it was happening before. The best way to illustrate the issue is with a screen capture:

![flicker](https://cloud.githubusercontent.com/assets/1239058/20485699/c0ead3d2-afca-11e6-9fe5-487dc6864925.gif)

All I'm doing in this video is typing, and I'm not saving the file or refreshing anything. Let me know if there is anything I can do to help diagnose the issue. Thanks!

**Comment by rchl** [comment]
Previously GitGutter was blocking the editor when updating markers so you wouldn't see the markers flickering but you would experience freezes in the editor instead.

I guess the bug is still valid. It can probably be fixed by delaying clearing of markers in some circumstances. But it won't necessarily be an easy fix.

**Comment by r-stein** [comment]
Can you try to remove/outcomment [this line](https://github.com/jisaacks/GitGutter/blob/1.4.0/git_gutter_show_diff.py#L68)?The regions are overwritten by readding them anyway so this should not be necessary.

**Comment by sgerace** [comment]
Sure thing, only I'm not sure where to make this change to have it take effect. I tried looking around for where Package Control installs the source code for the packages I have installed, but wasn't able to find it.

**Comment by r-stein** [comment]
Oh sorry install [PackageResourceViewer](https://packagecontrol.io/packages/PackageResourceViewer), press ctrl+shift+p and run the command "PackageResourceViewer: Open Resource". Aferwards select "GitGutter/git_gutter_show_diff.py" to open the file. Do the change and restart ST to be sure the changes are propagated.

**Comment by sgerace** [comment]
Yep, that seems to have fixed the issue.

**Comment by r-stein** [comment]
ok nice =)

**Comment by jisaacks** [comment]
What if you add a new line, causing the `+` to show up. Then remove the new line. Will the `+` still go away?

**Comment by sgerace** [comment]
Yes, it seems like it is behaving properly:

![flicker-2](https://cloud.githubusercontent.com/assets/1239058/20492337/563a23c2-afe2-11e6-9543-1333ddb4b2bb.gif)

**Comment by deathaxe** [comment]
I may have a deeper look on this and add it to pr #338 if you want.

**Comment by jisaacks** [comment]
@deathaxe I would make it a separate PR since it is a different bug.

**Comment by r-stein** [comment]
We should also ensure to only thread save call api functions on ST2.

**Comment by deathaxe** [comment]
Let me ask two questions as I never used ST2, please:

1. There is only one thread all plugins are running in, isn't it?
2. ST2 does not support *_async events and so there is only `sublime.set_timeout(...)` whose callback is called in the main thread anyway?
3. How can a plugin make API calls thread safe? I could create a mutex or so, but this won't prevent other plugins from calling the same api at the same time.

The only idea I have is to check, whether a Promise to update gutter is active and prevent running twice to avoid race conditions.

**Comment by rchl** [comment]
As you said, in ST2 `sublime.set_timeout` is used which calls on the main thread so there is no problems with thread safety.

In ST3, starting from git call, we run everything in secondary thread. To be super safe, we could go back on the main thread right after doing the git call by resolving the promise from `sublime.set_timeout` call. But not sure it's necessary as ST3 API is thread safe.

**Comment by deathaxe** [comment]
@jisaacks Ok, but I would like to base it on the last one, if possible.

This new pull request will include some bigger changes to git_gutter_show_diff.py as I found a signfificant performance issue with the protected regions handling while searching for a solution, yesterday.

It takes about 10 seconds to add gutter icons to 65k lines of code. After some first changes I could reduce that time to 0,2 seconds. But this needs some more work until ready to propose.

**Comment by jisaacks** [comment]
@deathaxe how close is #338 to being ready, perhaps we could get it merged in, and this wouldn't be an issue.

@rchl @r-stein @sigod what are your thoughts?

**Comment by rchl** [comment]
@jisaacks can we get the simpler pull requests merged first? Kinda hard to keep track when they are piling up and possibly are sort of dependent.

**Comment by jisaacks** [comment]
@rchl which simpler PR are you referring to?

**Comment by rchl** [comment]
Create only one instance of GitGutterSettings (fixes #334)
#337 opened 3 days ago by rchl  
 Some usage optimization
#331 opened 6 days ago by deathaxe • Approved  
Make "compare to" functionality work per git directory (fixes #327)
#328 opened 7 days ago by rchl  
 Fix error invoking git which happens on all of my Windows systems.
#262 opened on Dec 18, 2015 by jfultz • Approved

**Comment by jisaacks** [comment]
@rchl I went ahead and merged all those that you listed.

**Comment by deathaxe** [comment]
Nice!

**Comment by canoeberry** [comment]
Wondering why you haven't released this large set of useful fixes?

**Comment by canoeberry** [comment]
Sorry - very confused on the status of this work. I cannot see any commits, here OR in deathaxe's fork that claims to fix it. The last commit mentioned (073599c) doesn't appear anywhere that I can see. Very interested in seeing this fixed.

**Comment by deathaxe** [comment]
Those commits were on a throw away "preview" branch. Did realize the references here, too late.

The fix is actually ready on my local branch but as it is based on some pending PR, I am not able to publish it. Unfortunatelly the reviewers seem to have no time to verify the quality of those PR, so I can't merge them or go on publishing fixes as they want all PRs being merged by rebase.

**Comment by canoeberry** [comment]
Bummer. The flashing is really annoying ... on a good day ;-)

Thanks for getting back.

**Comment by deathaxe** [comment]
I periodically look for PRs being reviewed. But havn't been able to go on for about 2 month now. 

I tested everything with ST2 2221 and 3126 on Linux x64 and Windows x64. But I have no chance to check OS X compatibility. So I can't merge something without maybely breaking something on OS X. 

Merging not approved PRs is forbitten.

**Comment by canoeberry** [comment]
I'll test it on Mac OS X for you if you can send me a patch ...

**Comment by deathaxe** [comment]
You could pull the `preview` branch from https://github.com/deathaxe/gitgutter. It is a throw away preview branch with based on current master with all my open PR merged in. This may be good for testing all fixes/features in one run.

**Comment by canoeberry** [comment]
OK - testing it now. Thanks - I didn’t know which branch to use when I was poking around the other day.

Do you want some feedback? I am not a huge user. I only just discovered that I can hover over the gutter and see the actual diffs at that location and copy and/or revert and various other things.

**Comment by deathaxe** [comment]
I tried to explain the changes and/or new functions in each pull request. It would be helpful to know, wheter something throws exceptions or if you realize some strange behaviour. 

1. You could try to change the Gutter theme temporarily to "Default HIDPI.gitgutter-theme" to see whether the gutter icons are shown in double size correctly.
2. I added a command "Compare against File Commit" which lists all commits with changes to the open file.
3. You could try the new Quick Panel commands "Preferences: GitGutter ..."

**Comment by canoeberry** [comment]
Still seeing some blinking for sure. I am not yet sure how much blinking I am seeing though. I just wanted to let you know I was still playing with it.

**Comment by r-stein** [comment]
I am not completly sure how fixing the blinking is based on the pending PRs. I have just locally commented out that line (from https://github.com/jisaacks/GitGutter/issues/340#issuecomment-261960680) and no problems since then. However we may also ensure that the `_update_ui` is on the main thread or something like this.

**Comment by canoeberry** [comment]
Wow - this whole time I thought it was a sublime bug whereas it's actually a GitGutter one. I didn't read this issue carefully enough! I just checked out 1.3.0 and all is well again. I wish I had thought of that sooner.

I am going to figure out which commit caused this issue unless you already know which one it was.

**Comment by r-stein** [comment]
We moved from syncroned handling to a separate thread to improve the performance. Unfortunatly the UI callbacks are still on that thread and remove and readd those icons instead of being on the main thread or just replace them. So its not a single commit, but the whole background rewrite.
It's super easy to fix, but I didn't want to conflict with pending PRs and mainly support the diff-popup.

**Comment by canoeberry** [comment]
Well my guess is you already know the answer to this.

```
$ git bisect bad
d56c24ae1ebf548d5d9347911d5f84d0f29c4d24 is the first bad commit
commit d56c24ae1ebf548d5d9347911d5f84d0f29c4d24
Author: Rafal Chlodnicki <rchlodnicki@opera.com>
Date:   Thu Oct 27 22:38:06 2016 +0200

    Introduce async invocations of git (real non-blocking behavior) (#230)
    
    GitGutterHandler.run_command() now runs git in an async thread and returns a
    Promise which gets resolved when git operation finishes. For clarity of the
    code, rather than using hard to read callbacks, added simple implementation of
    Promise.
    
    The behavior is always non-blocking now (only in ST3), so removed "non_blocking"
    setting.

:100644 100644 6b6a20c9f016294857dddaad9885429ae7973f3b 9e658c4eda51b1c03cfb71f0067242dbe23341e6 M	GitGutter.sublime-settings
:100644 100644 2cc3edb8ba9219ee6b8b5bb513b2d5bbaf9535c1 c33e5d46eceb13a883e222fece0525c0ea46ad93 M	README.md
:100644 100644 2d72be433606c17f7e0cf551b24821ac7b2730e2 cbef8e2c34669183b4e3e8e3b5dc01424aecff16 M	git_gutter_compare.py
:100644 100644 fc8fb733f9f2ae0b0969b79b9fcec19101c6fd25 fb28059b2620f368a4365ef8913ee6663a2dc987 M	git_gutter_events.py
:100644 100644 50257162b5f344e929e230e2a861278e2eee7ddc c25e98c6f0124d240acc946b6c3abbc3e5243b2f M	git_gutter_handler.py
:100644 100644 9d839cc89a8dd437944f4ba519ca12ead78c8488 6c10126ef37dcbad0cd30dced5087602e7a5a2aa M	git_gutter_jump_to_changes.py
:100644 100644 b109fd5fe8dc12500ef230558c007955f2ea4ce0 2b71280130996de1923f2c72aa57e0063679bfe3 M	git_gutter_popup.py
:100644 100644 6716de4fb963979783ebe0ac0028d4b1093820cc ef0566731a6d277d15631a9f7761ba48386aeb65 M	git_gutter_show_diff.py
:000000 100644 0000000000000000000000000000000000000000 ec5050bb6b4476bdf508b39e1517af20eed9ea64 A	promise.py
:000000 040000 0000000000000000000000000000000000000000 d1114dc4e8cce12df3ff61d58c6469a8d7cf9049 A	tests
:000000 100644 0000000000000000000000000000000000000000 c2fdb3dcbb1d69f962032b12bdf7abb77a5495aa A	unittesting.json
[paradiddle.local:20:12:40:(no branch, bisect started on 311811e):~/src/SublimeText/GitGutter]
$ 
```

Obviously a move to async is ... a big deal ;-)

However, the solution must surely be to erase the markers in the same thread that you apply the new markers AFTER you get the results from the diff, right?

**Comment by r-stein** [comment]
I have created a PR #364 to fix the problem. You can checkout the branch and test it.

**Comment by canoeberry** [comment]
Thanks. Trying it now. It seems to work and makes complete sense. Still not sure why it caused flashing the old way. If I was just sitting there doing nothing, it still flashed. So what was the UI thread doing that would cause the erase to be done separate from the redraw? I even changed my cursor not to blink just in case that was the reason.

If you know the answer, I'd love to hear it for future reference.

BTW - This plugin is not unload/reload safe. Requires a restart of Sublime.

```
reloading plugin GitGutter.promise
unloading plugin GitGutter.view_collection
Traceback (most recent call last):
  File "/Users/jpayne/Applications/Sublime Text.app/Contents/MacOS/sublime_plugin.py", line 701, in is_enabled_
    ret = self.is_enabled()
  File "/Users/jpayne/Library/Application Support/Sublime Text 3/Packages/GitGutter/git_gutter.py", line 34, in is_enabled
    if not settings.git_binary_path:
AttributeError: 'Settings' object has no attribute 'git_binary_path'
Traceback (most recent call last):
  File "/Users/jpayne/Applications/Sublime Text.app/Contents/MacOS/sublime_plugin.py", line 699, in is_enabled_
    ret = self.is_enabled(**args)
  File "/Users/jpayne/Library/Application Support/Sublime Text 3/Packages/GitGutter/git_gutter.py", line 34, in is_enabled
    if not settings.git_binary_path:
AttributeError: 'Settings' object has no attribute 'git_binary_path'
Traceback (most recent call last):
  File "/Users/jpayne/Applications/Sublime Text.app/Contents/MacOS/sublime_plugin.py", line 699, in is_enabled_
    ret = self.is_enabled(**args)
  File "/Users/jpayne/Library/Application Support/Sublime Text 3/Packages/GitGutter/git_gutter.py", line 34, in is_enabled
    if not settings.git_binary_path:
AttributeError: 'Settings' object has no attribute 'git_binary_path'
Traceback (most recent call last):
  File "/Users/jpayne/Applications/Sublime Text.app/Contents/MacOS/sublime_plugin.py", line 701, in is_enabled_
    ret = self.is_enabled()
  File "/Users/jpayne/Library/Application Support/Sublime Text 3/Packages/GitGutter/git_gutter.py", line 34, in is_enabled
    if not settings.git_binary_path:
AttributeError: 'Settings' object has no attribute 'git_binary_path'
```
And perhaps other errors.

**Comment by r-stein** [comment]
Updating is always somewhat problematic with bigger packages, because it only reloads some modules. However that should not happen, because I didn't change that file or the settings file.

**Comment by r-stein** [comment]
> Thanks. Trying it now. It seems to work and makes complete sense. Still not sure why it caused flashing the old way. If I was just sitting there doing nothing, it still flashed. So what was the UI thread doing that would cause the erase to be done separate from the redraw? I even changed my cursor not to blink just in case that was the reason.

In the [`_update_ui`](https://github.com/jisaacks/GitGutter/blob/1.4.0/git_gutter_show_diff.py#L66) method it removes the regions in line 68 and readds them in line 69 (deleted), 70 (inserted), and 71 (modified). If this runs on the main thread the whole method is executed and the redraws are done afterwards. Hence there are no redraws. If it runs on a separate thread each api call is executed on its own. Therefore they are first removed and then added.

**Comment by canoeberry** [comment]
Maybe it's the version I was updating from (1.3.0).

Anyway - regarding the threads, I think what you're saying is, The UI thread is high priority. So, when you erase, the UI thread wakes up and performs the erase and goes back to sleep, and then the next step is performed, and the UI thread wakes up and does it again, etc.

Are python threads ... shoot what's the word ... getting old ... pre-emptive? Pre-emptive such that hgih priority threads will just run as soon as they become runnable?

**Comment by r-stein** [comment]
I don't exactly know how Sublime Text handles it threads. However all plugin commands are run on the same thread. From my observations I assume the UI updates run on the same thread. Therefore the UI updates are not possible during the method, because the thread is busy. If the calls are from an other thread they can immediately be executed, because the UI thread is idle.

**Comment by deathaxe** [comment]
> I am not completly sure how fixing the blinking is based on the pending PRs. I have just locally commented out that line (from #340 (comment)) and no problems since then. However we may also ensure that the _update_ui is on the main thread or something like this.

@r-stein: It is not fixed at all. Those PRs are the base to publish the fix. I need this PRs approved to publish the fix for Issue #340.

**Comment by deathaxe** [comment]
> We moved from syncroned handling to a separate thread to improve the performance. Unfortunatly the UI callbacks are still on that thread and remove and readd those icons instead of being on the main thread or just replace them.
@r-stein: The issue is, GitGutter removes all regions first and than calculates and adds the updated ones. This is what the user sees as flickering, because calculating updated regions takes 100ms to 10s depending on file size. I already prepeared a fix, but it is based on the mensioned PRs. Therefore it is not yet published. It is nearly impossible to keep everything together right now. Too many fixes in parallel which can't actually be handled totally seperatly.
